RFC6121 rosterver, XEP-0352 CSI and stream features extraction

rufferson opened the pull request:
This set of 3 patches 
1) adds ability to extract stream features - since not all of them related to wocky space.
In particular it extracts CSI, AMP, rosterver and SM features flagging them on connection feature mask.
2) extends google:queue functionality to support client state indication (CSI) as per XEP-0352
3) adds implementation of the roster versioning and consequently roster cache. Roster is cached in sqlite3 db and cache is serialized and injected as WockyNode tree  into roster processing.

Kaffeine commented:
Hi there. I can't review non-trivial patches and it would be unresponsible to just merge PRs.
@rufferson and @noonien-d You both seem to know the code and it is a real chance to move forward. I think it would be acceptable if you'll review each-other PRs (and I merge them).

@noonien-d can you please review this series?

rufferson commented:
Rebased on master's tip, fixed styling (to pass make check) and added some make-up to make it build under GCC9 (Arch). Note - still doesn't include new wocky api

noonien-d commented:
I'll review the next days.
@rufferson would you mind to add an additional PR to enable the wocky api changes?

rufferson commented:
I'd expect it to be bundled with commit updating the submodule pointer, otherwise the change would breaking and will need another commit to recover. I was thinking today adding conditional interface change but even that is non reliable as commit which removes peername from callback does not change any macro definitions.

Kaffeine commented:
Do you have a version definition in wocky? Like `OPENSSL_VERSION_NUMBER`?
It is a typical mean to introduce changes (especially backward incompatible).
We had to do the same in telepathy-nonsense: https://github.com/TelepathyIM/telepathy-nonsense/blob/master/connection.cc#L242

rufferson commented:
Unfortunatelly no, it was carved out from gabble without that, currently version is statically defined in wocky/wocky.pc as 0.0.0.1

noonien-d commented on configure.ac:
```diff
@@ -220,7 +221,7 @@ PKG_CHECK_MODULES(GLIB,
 
 AC_DEFINE(GLIB_VERSION_MIN_REQUIRED, GLIB_VERSION_2_32, [Ignore post 2.32 deprecations])
-AC_DEFINE(GLIB_VERSION_MAX_ALLOWED, GLIB_VERSION_2_32, [Prevent post 2.32 APIs])
+AC_DEFINE(GLIB_VERSION_MAX_ALLOWED, GLIB_VERSION_2_34, [Prevent post 2.32 APIs])
```
Minor: description should be "Prevent post 2.32 APIs"

noonien-d commented on src/capabilities.c:
```diff
@@ -96,10 +96,12 @@ static const Feature self_advertised_features[] =
   { 0, NULL }
 };
 
+#if 0
```
Remove this if its not needed?

noonien-d commented on src/roster-cache.c:
```diff
@@ -0,0 +1,748 @@
+  const gchar *dir, *path;
+
+  /* This should still work if it's the magic value ":memory:". */
+  path = g_getenv ("ROSTER_CACHE");
```
where does `ROSTER_CACHE` come from?

noonien-d commented on src/roster.c:
```diff
@@ -1826,13 +1870,28 @@ connection_status_changed_cb (GabbleConnection *conn,
+                      wocky_node_free (query);
+                    }
+                }
               stanza = _gabble_roster_message_new (self, WOCKY_STANZA_SUB_TYPE_GET,
```
Shouldn't you avoid requesting the roster after you loaded the cache successfully? Do I miss something?

rufferson replied:
No, we need to request roster updates since last cached version. Cache allows us to save the bandwidth as without any changes we'd get simple <iq type=result/> as confirmation. Otherwise we have two possibilities - server floods updates at us or send full roster (up to server to decide what's better)

noonien-d replied:
I see, thanks for the explanation.

noonien-d commented on src/roster.c:
```diff
@@ -1410,7 +1445,13 @@ got_roster_iq (GabbleRoster *roster,
+      process_roster (roster, query_node);
+      user = conn_util_get_bare_self_jid (priv->conn);
+      DEBUG ("updating roster cache for %s", user);
+      roster_cache_update_roster (roster->priv->rcache, user, query_node);
```
Needs to check if rcache != NULL, in case the server doesn't support rosterver.

noonien-d commented on src/roster-cache.h:
```diff
@@ -0,0 +1,82 @@
+gboolean roster_cache_update_roster (RosterCache *,
+    const gchar *, WockyNode *);
+
+RosterCache *
```
No need to expose roster_cache_new

noonien-d commented on plugins/console/plugin.c:
```diff
@@ -76,6 +76,7 @@ plugin_iface_init (
 
   iface->name = "XMPP console";
   iface->version = PACKAGE_VERSION;
+  iface->sidecar_interfaces = sidecar_interfaces;
```
I'm not sure about this, why did you add it?

rufferson replied:
this eventually a fix for this sidecar plugin as the reason the static variable was unused is because this line was missing. The sidecar though is rather example so no one ever cared I presume whether it actually worked. I can probably move it to separate commit as there's also related python script (client side of the sidecar interface - to test its functionality) which also needs similar one line fix to make it work and the commit title doesn't match the purpose

rufferson commented:
Thanks for review, that was really helpful. 
I cannot think of any sensible way to handle it except to schedule the retry for later which seems overly complicated. Even if cache is stuck - it's non-breaking issue, each login server will be sending the diff, after the diff grows too big server will simply start sending entire roster (ignoring the cache). What I will do perhaps is just make it visible (Err-out, currently it's debug from cache itself) so that someone can take a look and either fix or drop it.

rufferson commented:
btw I see you have implemented CSI in wocky, why didn't you flag it here? I think it makes more sense having it in wocky on top of SM resumption queue. And I've pushed it here just because back then I considered wocky to be overly complex :) I'll gladly ditch it in favour of reviving SM/CSI wocky branch

noonien-d commented:
Actually I wondered if its not nicer to keep in in gabble, if possible. Can you explain why there is the need to keep it with sm?

rufferson commented:
My line of thinking is that on mobile devices we want to minimize radio usage, so if we got powersave indication and flushed the queue - we do not want to do any residual writes anymore to avoid waking/inhibiting sleep of the radio accidentally. Therefore after we sent inactive we don't write but queue instead, and on active - flush again.

noonien-d commented:
Why don't we enable power saving mode in gabble when google queue is available?

rufferson commented:
well... yes, actually one (csi) could be independent from another (wocky queue)... so we may make it configurable to enforce queue in addition to indication for mobile platforms. Ok, convinced :)

Kaffeine commented:
@rufferson I'm ready to merge this once you confirm it.

rufferson commented:
was thinking to add gcc10 cleanups here but will probably rather push them with next pr for carbons management. So yes, good to go. Thanks for review.
